src/toil/spark/spark_cluster.py

```diff
@@ -0,0 +1,328 @@
+    except OSError:
+        raise RuntimeError('docker not found on system. Install on all nodes.')
+
+
+class SparkService(Job.Service):
+
+    def __init__(self,
+                 sudo,
+                 memory,
+                 numWorkers,
+                 cores = None):
+
+        self.sudo = sudo
+        self.memory = memory
+
+        if cores:
+            self.cores = cores
+        else:
+            self.cores = multiprocessing.cpu_count()
+
+        print >> sys.stderr, "Starting master"
+            
+        self.numWorkers = numWorkers
+        Job.Service.__init__(self, memory = self.memory, cores = self.cores)
+
+
+    def _add_worker_services(self):
+        """
+        Starts the worker services.
+        """
+        _log.info("Starting %d Spark workers and HDFS Datanodes.", self.numWorkers)
+        
+        for i in range(self.numWorkers):
+            self.addChild(WorkerService(self.IP, self.sudo, "%s G" % self.memory))
+            
+
+    def start(self, fileStore):
+        """
+        Start spark and hdfs master containers
+
+        fileStore: Unused
+        """
+        
+        self.IP = subprocess.check_output(["hostname", "-f",])[:-1]
+
+        _log.info("Started Spark master container.")
+        self.sparkContainerID = _docker_call(no_rm = True,
+                                             work_dir = os.getcwd(),
+                                             tool = "quay.io/ucsc_cgl/apache-spark-master:1.5.2",
+                                             docker_parameters = ["--net=host",
+                                                                  "-d",
+                                                                  "-v", "/mnt/ephemeral/:/ephemeral/:rw",
+                                                                  "-e", "SPARK_MASTER_IP="+self.IP,
+                                                                  "-e", "SPARK_LOCAL_DIRS=/ephemeral/spark/local",
+                                                                  "-e", "SPARK_WORKER_DIR=/ephemeral/spark/work"],
+                                             tool_parameters = [],
+                                             sudo = self.sudo,
+                                             check_output = True)[:-1]
+        _log.info("Started HDFS Datanode.")
+        self.hdfsContainerID = _docker_call(no_rm = True,
+                                            work_dir = os.getcwd(),
+                                            tool = "quay.io/ucsc_cgl/apache-hadoop-master:2.6.2",
+                                            docker_parameters = ["--net=host",
+                                                                 "-d"],
+                                            tool_parameters = [self.IP],
+                                            sudo = self.sudo,
+                                            check_output = True)[:-1]
+
+        # attach worker service jobs
+        self._add_worker_services()
```

Comment:
What's the reason for that? Ideally, we would start the worker services after the HDFS namenode is up. Otherwise, this makes a race condition worse...